Color allowed and actual overproduction

veger opened the pull request:
First I cleaned up the coloring logic. Unfortunately the link flags are not documented, so I have no idea what some of the flags mean (in their nuanced ways)... Especially `ChildNotMatched` I have no idea about.

I added a new color for allowed overproduction. I picked magenta as it is close to red (error), feel free to provide another color scheme. BTW For light/dark theme I switched the main/highlight colors (I am not too artistic and finding only/just one set of colors was already the hardest part of the PR :rofl: )

Activated Carbon allows for overproduction:
![image](https://github.com/have-fun-was-taken/yafc-ce/assets/171827/51954692-4de5-4e35-81ee-174a5fedec19)

Activated Carbon is over produced (forced by the increased Nitrogen production):
![image](https://github.com/have-fun-was-taken/yafc-ce/assets/171827/713d1882-ad50-461f-9739-86772b960557)

The allowed overproduction is also shown in the Summary tab (actual over production is not, as there are already enough errors when the production sheets are unbalanced).

Fixes #8

have-fun-was-taken commented:
Could you please elaborate on why an item being overproduced has red color? It is the same color as the warning that there is an error in the solution.

veger commented:
I picked red, as it was already used in the top to show overproduced products.
![image](https://github.com/have-fun-was-taken/yafc-ce/assets/171827/45f3838c-3188-41d4-9414-16ca45178985)

And basically it actually _is_ an error, as it will clog the factory to an halt given time.

But I am fine to change colors. I mentioned somewhere else as well, the colors are not very clear to me to begin with. So we could reconsider them, or at the least implement some legend (info/documentation screen or popup) to clarify them to the users.

have-fun-was-taken commented:
Before anything else, I'd like to thank you for your continued effort and the new features that you bring to YAFC.

> I picked red, as it was already used in the top to show overproduced products.

To clarify the situation, I would like to refer to the [issue](https://github.com/have-fun-was-taken/yafc-ce/issues/8) that this PR aims to solve. In short, the issue says that it would be good to see when something is overproduced, so there are no sudden deadlocks.

For that, I suggest to color a product in purple only if it's actually overproduced. Otherwise to keep its color as usual. The reason for this is manifold.

1. One color keeps the change unintrusive when bringing the functionality to fix the issue. That allows us to not burden the UI with more color-rules to remember unless it's necessary.

2. We prevent the confusion. If a linked _byproduct_ is overproduced, it does not color it red, preventing the confusion of whether it's a solution-error or not. It colors it purple, signifying that it's overproduced.

3. We set one color responsible for the concept. If something is actually overproduced, then it's colored in purple.

Here's a real example of how my points play out. Please find attached the [project file](https://files.catbox.moe/eqzbvu.yafc) with the page `Nuclear powergen`. 
The current UI shows the following:

![image](https://github.com/have-fun-was-taken/yafc-ce/assets/73699379/9d136ac6-98b5-468e-96c6-aceef21c17dc)

Electricity is overproduced. It is red in the desired products, but it's shown as just linked in the rest of the sheet. The linked Plutonium-239 1.81/s in byproducts is overproduced but is showed only as linked throughout the whole sheet.

The current PR's coloring is as follows:

![image](https://github.com/have-fun-was-taken/yafc-ce/assets/73699379/14b1fef3-9a69-4034-9c21-c6d52874693d)

Plutonium-238 is shown as linked in desired products but has purple background on the sheet, which is confusing. Electricity is red in desired and byproducts, but shown as just linked on the sheet. And Plutonium-239 is shown as red everywhere.
That many colors in different contexts is bad for the UI because it adds to the visual load of the user, making the UI less intuitive and more frustrating to use for a new user.

I suggest the following scheme:

![image](https://github.com/have-fun-was-taken/yafc-ce/assets/73699379/b9418e74-4700-49c8-8851-bc6de1205ee8)

Note that Plutonium-238 is shown as just linked even though the overproduction is allowed but doesn't happen. In other words, I suggest that if something is actually overproduced, we indicate it to the user in no uncertain terms. The purple color will be responsible for that. If something is actually overproduced, then it's purple. If it's not overproduced now, then it's invisible and doesn't spend the attention of the user.

Please let me know what you think.

have-fun-was-taken commented:
Here's the [savefile](https://files.catbox.moe/k6gw3y.zip) to sync the mods if you want to load the yafc [project file](https://files.catbox.moe/eqzbvu.yafc) that I linked above.

veger commented:
Basically it boils it down to

> We set one color responsible for the concept. If something is actually overproduced, then it's colored in purple

It is nice and clear/simple! :smile: 

I read the issue as "please show what is _allowed_ to be overproduced". When I implemented this, I thought it that it would be more convenient to have the _actual_ overproduction colored as well. (Maybe I misread/misunderstood the issue?)

But as said, I like this simple color scheme much better. When I have some time (maybe tomorrow otherwise most likely on Friday) I'll adjust this PR accordingly.

have-fun-was-taken commented:
To not review work-in-progress, please let us know when the PR is ready for the review.

veger commented:
I adjusted the PR to
* Only show actual overproduction
* Use magenta (not Error/red)
* Include the overproduction indicator color for the 'Desired products and amounts' icons

A slight changed compared to the suggested scheme: over production indicator is only showed at the production icons (not at the ingredients icons) of the recipe. As I felt that it is a production issue and show now show in (all) recipes that simply use the over produced item.

Also, I refactored the logic the determine the icon background color for the SummaryView before these changes. I decided to keep it, as it is much more clear than the single liner it was before (even there are no functional changes in the SummaryView naymore).

have-fun-was-taken commented:
Hmm, I saw the green color and considered that the changes were not over yet. I'll take another look.

veger commented:
Besides the change to the schema I described (and felt it made/makes sense), all should be as discussed/proposed now.

have-fun-was-taken commented on YAFCui/Core/Structs.cs:
```diff
@@ -35,6 +35,11 @@ public enum SchemeColor {
         GreyText,
         GreyTextFaint,
+        // Magenta group (indicate overproduction)
+        Magenta,
```
Out of the four new entries, only `Magenta` is used. Do we need the other three in there?

veger replied:
Yes, the 4 entries are used for 'default', 'hover', 'text', 'disabled text' or something.

See https://github.com/have-fun-was-taken/yafc-ce/pull/70/files#diff-4de221ae69c7d3ff01036e4865ccd3ef4b7541e70c739e28e54ad1b0f686decaR35 for actual/correct details :smile:

veger replied:
Technically, it might be the case that those enums do not to be defined expliitly (maybe :thinking: ), but for the other colors the same is done. So each color is always in a group of four 'sub colors'.

have-fun-was-taken replied:
Thank you for the explanation!

have-fun-was-taken commented:
All of these `elseif`s, comments for each case, and long conditions make me itching for a refactor, but that's a story for another day.

veger commented:
I refactored from a single line, to the if-else structure. It clarifies a lot already. How to do further I do not know, AFAIK it is not allowed in C# to do a
```
switch true {
case condition1:
  break;
case condition2:
  break;
default:
}
```
?

have-fun-was-taken commented:
I like the approach of 
```
public void foo(input) {

boolean longConditionExplainedByItsName = longBooleanExpression;
if (longConditionExplainedByItsName) {
action1();
return;
}

boolean longCondition2 = ...
if(longCondition2) {
action2();
return;
}

}
```
It makes the logic branch less, improving the speed of understanding.

have-fun-was-taken commented:
> I refactored from a single line, to the if-else structure. It clarifies a lot already.

I agree that the version in your PR is much more readable than the oneliners it was before.

veger commented:
If you want I can wrap the color logic into functions and return the correct color?

The `longConditionExplainedByItsName` got (already) implemented by the comments I left on each `if` branch.

have-fun-was-taken commented:
> If you want I can wrap

This idea is just my understanding of readability, and it will need to be tested, so I don't ask for any changes from your side in this regard.

have-fun-was-taken reviewed (approved):
Checked the changes on the `Nuclear powergen` example that was used above. The color scheme is as described.

veger commented:
I'll merge to unblock @shihan42 in #59
